src/Interpreters/MonotonousOrderByVisitor.h

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <DataTypes/DataTypeFactory.h>
+#include <Functions/FunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/IAST.h>
+#include <Common/typeid_cast.h>
+
+namespace DB
+{
+
+using Monotonicity = IFunctionBase::Monotonicity;
+
+class MonotonicityCheckMatcher
+{
+public:
+    struct Data
+    {
+        const TablesWithColumns & tables;
+        const Context & context;
+        ASTPtr & identifier;
+        Monotonicity & monotonicity;
+        DataTypePtr data_type = nullptr;
+        bool first = true;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        auto * ast_function = ast->as<ASTFunction>();
+        if (ast_function)
+        {
+            visit(*ast_function, data);
+        }
+    }
+
+
+    static void visit(ASTFunction & ast_function, Data & data)
+    {
+        if (!data.monotonicity.is_always_monotonic)
+            return;
+
+        auto arguments = ast_function.arguments;
+        if (arguments->children.size() != 1)
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        if (data.first)
+        {
+            data.identifier = ast_function.arguments->children[0];
+            if (!data.identifier->as<ASTIdentifier>())
+            {
+                data.monotonicity.is_always_monotonic = false;
+                return;
+            }
+        }
+
+        auto * identifier_ptr = data.identifier->as<ASTIdentifier>();
+
+        if (AggregateFunctionFactory::instance().isAggregateFunctionName(ast_function.name))
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        const auto & function = FunctionFactory::instance().tryGet(ast_function.name, data.context);
+        auto pos = IdentifierSemantic::getMembership(*identifier_ptr->as<ASTIdentifier>());
+        if (pos == NULL)
```

Comment:
Do not use `NULL` in C++ at all. It's C legacy. You need `if (!pos)` here and `(x == nullptr)` in general C++ cases.